packages/cli/src/util/compile-vercel-config.ts

```diff
@@ -0,0 +1,132 @@
+import { existsSync } from 'fs';
+import { mkdir, writeFile, unlink } from 'fs/promises';
+import { join } from 'path';
+import output from '../output-manager';
+import { NowBuildError } from '@vercel/build-utils';
+import { VERCEL_DIR } from './projects/link';
+import { ConflictingConfigFiles } from './errors-ts';
+
+export interface CompileConfigResult {
+  configPath: string | null;
+  wasCompiled: boolean;
+}
+
+export async function compileVercelConfig(
+  workPath: string
+): Promise<CompileConfigResult> {
+  const vercelJsonPath = join(workPath, 'vercel.json');
+  const nowJsonPath = join(workPath, 'now.json');
+  const hasVercelJson = existsSync(vercelJsonPath);
+  const hasNowJson = existsSync(nowJsonPath);
+
+  // Check for conflicting vercel.json and now.json
+  if (hasVercelJson && hasNowJson) {
+    throw new ConflictingConfigFiles([vercelJsonPath, nowJsonPath]);
+  }
+
+  // Only check for vercel.ts if feature flag is enabled
+  if (!process.env.VERCEL_TS_CONFIG_ENABLED) {
+    return {
+      configPath: hasVercelJson
+        ? vercelJsonPath
+        : hasNowJson
+          ? nowJsonPath
+          : null,
+      wasCompiled: false,
+    };
+  }
+
+  const vercelTsPath = join(workPath, 'vercel.ts');
+  const vercelDir = join(workPath, VERCEL_DIR);
+  const compiledConfigPath = join(vercelDir, 'vercel.json');
+
+  const hasVercelTs = existsSync(vercelTsPath);
+
+  if (hasVercelTs && hasNowJson) {
+    throw new ConflictingConfigFiles(
+      [vercelTsPath, nowJsonPath],
+      'Both vercel.ts and now.json exist in your project. Please use only one configuration method.',
+      'https://vercel.com/docs/projects/project-configuration'
+    );
+  }
+
+  if (hasVercelTs && hasVercelJson) {
+    throw new ConflictingConfigFiles(
+      [vercelTsPath, vercelJsonPath],
+      'Both vercel.ts and vercel.json exist in your project. Please use only one configuration method.',
+      'https://vercel.com/docs/projects/project-configuration'
+    );
+  }
+
+  if (!hasVercelTs) {
+    return {
+      configPath: hasVercelJson
+        ? vercelJsonPath
+        : hasNowJson
+          ? nowJsonPath
+          : null,
+      wasCompiled: false,
+    };
+  }
+
+  const tempOutPath = join(vercelDir, 'vercel-temp.js');
+
+  try {
+    const { build } = await import('esbuild');
+
+    await mkdir(vercelDir, { recursive: true });
+
+    await build({
+      entryPoints: [vercelTsPath],
+      bundle: true,
+      platform: 'node',
+      format: 'cjs',
+      outfile: tempOutPath,
+      packages: 'external',
+      target: 'node18',
+      sourcemap: 'inline',
+    });
+
+    delete require.cache[require.resolve(tempOutPath)];
+    const configModule = require(tempOutPath);
+    const config = configModule.default || configModule.config || configModule;
```

Comment:
Will you be adding any validation to the output?

Reply:
yes, the output will be validated automatically the same way vercel.json is in the commands: dev, build, deploy, where this util will be called.